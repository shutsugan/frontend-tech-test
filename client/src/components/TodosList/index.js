import React from 'react';

import { useStoreState } from 'easy-peasy';

const TodoList = _ => {
    const todos = useStoreState(state => state.todos.list);
    
    return (
        <div className="todo-list">
            todo list
        </div>
    );
};

export default TodoList;