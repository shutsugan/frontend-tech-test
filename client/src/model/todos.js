import { action, computed, thunk } from "easy-peasy";

const base_url = 'http://localhost:9001';

const noop = _ => {};
export const save = async (axios, actions, method, url, callback = noop) => {
    const res = await axios[method](`${base_url}/task/${url}`);
    const { task, message } = await res.data;

    callback(task);
    actions.setMessage(message);
};

export const checkProps = (actions, title, description) => {
    if (!title && !description) return actions.setMessage('Fields are reuqired');
};

export default {
    todos: [],
    selectedTodo: null,
    message: null,

    list: computed(state => Object.values(state.todos)),

    //actions
    setMessage: action((state, message) => state.message = message),
    selectTodo: action((state, todo) => state.selectedTodo = todo),
    added: action((state, todo) => state.todos[todo.id] = todo),

    fetched: action((state, payload = []) => {
        state.todos = payload.reduce((acc, todo) => {
            acc[todo.id] = todo;

            return acc;
        }, {});
    }),

    //thunks
    fetchTodos: thunk(async (actions, _, { injections }) => {
        const { axios } = injections
        const res = await axios.get(`${base_url}/tasks`);
        const { tasks } = await res.data;

        actions.fetched(tasks);
    }),

    addTodo: thunk(async (actions, { title, description }, { injections }) => {
        const { axios } = injections

        checkProps(actions, title, description);
        save(
            axios,
            actions,
            'post',
            `create/${title}/${description}`,
            todo => actions.added(todo)
        );
    }),

    updateTodo: thunk(async (actions, { id, title, description }, { injections }) => {
        const { axios } = injections;

        checkProps(actions, title, description);
        save(
            axios,
            actions,
            'put',
            `update/${id}/${title}/${description}`,
            _ => {
                actions.fetchTodos();
                actions.selectTodo(null);
            }
        );
    }),

    removeTodo: thunk(async (actions, id, { injections }) => {
        const { axios } = injections;

        save(
            axios,
            actions,
            'delete',
            `delete/${id}`,
            _ => actions.fetchTodos()
        );
    })
};