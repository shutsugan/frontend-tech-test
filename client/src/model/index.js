import todos from './todos';
import { thunk } from 'easy-peasy';

export default {
    todos,
    initialiseTodos: thunk(async (actions, payload, { dispatch }) => {
        await dispatch.todos.fetchTodos();
    })
};