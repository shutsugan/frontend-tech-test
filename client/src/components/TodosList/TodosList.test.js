import React from 'react';
import renderer from 'react-test-renderer';
import { createStore, StoreProvider } from 'easy-peasy';
import { expect as chaiExpect } from 'chai';

import model from '../../model';
import TodosList from './index';

const store = createStore(
    model, { injections: { axios: jest.fn() } }
);

describe('TodosList component', () => {
    it ('should render todos list component', () => {
        const tree = renderer.create(
            <StoreProvider store={store}>
                <TodosList />
            </StoreProvider>
        ).toJSON();
        
        expect(tree).toMatchSnapshot();
        chaiExpect(tree).to.be.an('object');
        chaiExpect(tree).to.have.own.property('props');
    });
})