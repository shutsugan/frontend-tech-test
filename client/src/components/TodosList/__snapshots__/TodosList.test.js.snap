// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodosList component should render todos list component 1`] = `
<div
  className="todo-list card full flex-column mrb-16"
/>
`;
